tests/interlocked_int/interlocked_int.c

```diff
@@ -0,0 +1,855 @@
+Tests_SRS_INTERLOCKED_43_042: [ interlocked_compare_exchange_pointer shall return the initial value of *destination. ]
+*/
+
+TEST_FUNCTION(interlocked_compare_exchange_pointer_equal_case)
+{
+    ///arrange
+    int value1 = 1;
+    int value2 = 2;
+    volatile void* destination = &value1;
+    void* comperand = &value1;
+    void* exchange = &value2;
+
+    ///act
+    void* return_val = interlocked_compare_exchange_pointer(&destination, exchange, comperand);
+
+    ///assert
+    ASSERT_ARE_EQUAL(void_ptr, &value2, destination);
+    ASSERT_ARE_EQUAL(void_ptr, &value1, return_val);
+}
+/*
+Tests_SRS_INTERLOCKED_43_010 [interlocked_compare_exchange_pointer shall compare the pointers destination and comperand. If they are equal, *destination is set to exchange. These operations are performed atomically.]
+Tests_SRS_INTERLOCKED_43_042: [ interlocked_compare_exchange_pointer shall return the initial value of *destination. ]
+*/
+TEST_FUNCTION(interlocked_compare_exchange_pointer_not_equal_case)
+{
+    ///arrange
+    int value1 = 1;
+    int value2 = 2;
+    volatile void* destination = &value1;
+    void* comperand = &value2;
+    void* exchange = &value2;
+
+    ///act
+    void* return_val = interlocked_compare_exchange_pointer(&destination, exchange, comperand);
+
+    ///assert
+    ASSERT_ARE_EQUAL(void_ptr, &value1, destination);
+    ASSERT_ARE_EQUAL(void_ptr, &value1, return_val);
+}
+
+/*
+Tests_SRS_INTERLOCKED_43_011[ interlocked_decrement shall atomically decrement (decrease by one) the 32-bit variable *addend.]
+Tests_SRS_INTERLOCKED_43_043: [ interlocked_decrement shall return the decremented value. ]
+*/
+TEST_FUNCTION(interlocked_decrement_upper_bound)
+{
+    ///arrange
+    volatile int32_t addend = INT32_MAX;
+
+    ///act
+    int32_t return_val = interlocked_decrement(&addend);
+
+    ///assert
+    ASSERT_ARE_EQUAL(int32_t, INT32_MAX - 1, addend);
+    ASSERT_ARE_EQUAL(int32_t, INT32_MAX - 1, return_val);
+}
+/*
+Tests_SRS_INTERLOCKED_43_011[ interlocked_decrement shall atomically decrement (decrease by one) the 32-bit variable *addend.]
+Tests_SRS_INTERLOCKED_43_043: [ interlocked_decrement shall return the decremented value. ]
+*/
+TEST_FUNCTION(interlocked_decrement_lower_bound)
+{
+    ///arrange
+    volatile int32_t addend = INT32_MIN + 1;
+
+    ///act
+    int32_t return_val = interlocked_decrement(&addend);
+
+    ///assert
+    ASSERT_ARE_EQUAL(int32_t, INT32_MIN, addend);
+    ASSERT_ARE_EQUAL(int32_t, INT32_MIN, return_val);
+
+}
+
+/*
+Tests_SRS_INTERLOCKED_43_012[ interlocked_decrement_16 shall atomically decrement (decrease by one) the 16-bit variable *addend.]
+Tests_SRS_INTERLOCKED_43_044: [ interlocked_decrement_16 shall return the decremented value. ]
+*/
+TEST_FUNCTION(interlocked_decrement_16_upper_bound)
```

Comment:
>interlocked_decrement_16_upper_bound [](start = 14, length = 36)

carefully chosing the bounds to avoid a underflow/overflow. but those will happen for real.

I believe it is better to have those cases exposed intentionally. That is a INTn_MAX that is incrementd and INTn_MIN that is decremented. Because sweeping under the rug maybe is a not a good idea. Should be explicit about the spiky cases.